### PR TITLE
Implement SSD Preprocessor and Training Script

### PR DESCRIPTION
**Pull request by farazali7**
This PR is for connecting the individual modules (PGM, SDM) of the WSL approach together for a training run. Alongside the development of a data pipeline for preprocessing and training the model, one notable change to the structure comes in the form of a Transfer Block, which sits between the CPM and DPM. The transfer block is defined more in detail in Zhang et. al  (2018)'s [paper on RefineDet](https://arxiv.org/pdf/1711.06897.pdf). The overall structure is shown below:

image

The transfer block takes input from our CPM, and propagates downstream feature maps while deconvolving them (DPM) to retain large-scale context.

**Summary of changes:**
- Added Preprocessor for loading, parsing, augmenting, and batching the dataset
- Created `train.py` script for orchestrating a training run of the model
- Introduced new Detection and Transfer blocks for better code structure
- Edits to loss function
- Edits to encoding and decoding functions

**Pseudo BBox Extraction Algorithm:**
The next phase of the project is experimentation. This requires us to have a good configuration of the bbox extraction algorithm. Please check out `extract_bbox_from_heatmaps.py` for the code pertaining to this critical part of the model architecture. There you will see the different parameters that can be configured. Post-processing the extracted bboxes using something like the contour size could also prove helpful. Below I have attached a few combinations of different parameter settings, showing how each changes the resulting bounding boxes. There is a balance between adding signal vs. adding noise which needs to be scrutinized a bit more to find decent algorithm settings.

Threshold = 127:
017FD5~1

Threshold = 127 & Dilation with kernel (1,5):
017FD5~3

Threshold = 127 & Dilation with kernel (1,10):
010DC4~1

Otsu's Thresholding:
017FD5~2

Otsu's Thresholding & Dilation with kernel (1,5):
017FD5~4

Otsu's Thresholding & Dilation with kernel (1,10):
01C1AA~1

### CODE REVIEW

File: src/detection/utils/ssd_utils/extract_bbox_from_heatmap.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 def extract_bbox_from_heatmap(heatmap, threshold=None, max_val=255, thresh_type=cv2.THRESH_BINARY, is_grayscale=True,
                               perform_otsu=False, perform_dilation=False, dilation_kernel_size=(1, 5), dilation_iters=1,
-                              min_contour_size=20):
+                              min_contour_size=0):
```
Comment (hmathlee) [inline src/detection/utils/ssd_utils/extract_bbox_from_heatmap.py:7]: Could we put this in the config file?

File: src/detection/utils/ssd_utils/output_encoder.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    :return: Encoded classification and localization tensors
+    '''
+
+    batch_size = boxes_batch.shape[0]
+    num_default_boxes = default_boxes.shape[-2]
+    ious = iou(boxes_batch, default_boxes)
+
+    def encode_at_batch_idx(idx, full_cls_tensor, full_loc_tensor):
+        '''
+        Helper function used for encoding (matching) GT boxes and default boxes at certain idx of a batch.
```
Comment (hmathlee) [inline src/detection/utils/ssd_utils/output_encoder.py:26]: It might help to add docs for the parameters here.

**Review by hmathlee — COMMENTED** [review]
Amazing work as always, Faraz. I don't have any comments except for a few documentation things here and there

File: requirements.txt [hunk 3]
```diff
@@ -1,7 +1,7 @@
 tensorboard
 tensorflow==2.5.1
 tensorflow-addons==0.12.1
-opencv_python==4.2.0.34
+opencv_python
```
[the remark this answers is not in the document]
Comment (farazali7) [reply requirements.txt]: Good point, I added the specific version in the reqs file now since it'll help with reproducibility.

File: src/detection/utils/ssd_utils/extract_bbox_from_heatmap.py (hunk 1 above)
Comment (farazali7) [reply src/detection/utils/ssd_utils/extract_bbox_from_heatmap.py]: That's a fantastic idea, I've put the configurable parameters of this function in the config file. Cleaner to use now in other parts of the code 🙂 .

File: src/explainability/gradcam.py [hunk 4]
```diff
@@ -100,11 +103,29 @@ def get_gradcam_plusplus_heatmap(model, img):
 
         return heatmap
 
-    def apply_gradcam(self, frame_df):
+    @staticmethod
+    def get_gradcam_heatmap(img, img_dim, model_base, last_conv_layer, heatmap_gen_fn, gt_label_idx=None):
```
[the remark this answers is not in the document]
Comment (farazali7) [reply src/explainability/gradcam.py:107]: Yep, added in latest commit.

File: src/detection/utils/ssd_utils/output_encoder.py (hunk 2 above)
Comment (farazali7) [reply src/detection/utils/ssd_utils/output_encoder.py:26]: Added in latest commit.

File: src/detection/custom_layers/detection_block.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras.layers import Conv2D
+from src.detection.utils.ssd_utils import get_default_box_count
+from src.detection.utils.bbox_utils import clip_boxes
+
+
+class DetectionBlock(tf.keras.Model):
+    def __init__(self, num_classes, aspect_ratios, extra_box_for_ar1, scale, image_size, offset, clip_boxes=False,
+                 kernel_regularizer=tf.keras.regularizers.l2(5e-4), **kwargs):
```
Comment (bvanberl) [inline src/detection/custom_layers/detection_block.py:10]: Is the default value for L2 described in the paper?
Also, might be a good idea to initialize the object in the function itself. But you could take the L2 hyperparameter as an optional parameter with default value.

File: src/detection/metrics/metrics.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
+import numpy as np
+import tensorflow as tf
+from tqdm import tqdm
+
+from src.detection.utils.bbox_utils import iou
+
+
+class MeanAveragePrecision(tf.keras.metrics.Metric):
```
Comment (bvanberl) [inline src/detection/metrics/metrics.py:8]: I absolutely love that you implemented this yourself -- what a feat!!
Did you see that TF has an implementation of this metric? https://www.tensorflow.org/ranking/api_docs/python/tfr/keras/metrics/MeanAveragePrecisionMetric

It's implemented in some TF  ranking module https://www.tensorflow.org/ranking/api_docs/python/tfr

**Review by bvanberl — APPROVED** [review]
Extraordinary work Faraz!

File: src/detection/custom_layers/detection_block.py (hunk 5 above)
Comment (farazali7) [reply src/detection/custom_layers/detection_block.py:10]: Yeah good idea, changed this parameter to just be the (float) regularization penalty and have object instantiate inside constructor. Also, paper did not describe a default value so I used 5e-4 as a sensible default.

File: src/detection/metrics/metrics.py (hunk 6 above)
Comment (farazali7) [reply src/detection/metrics/metrics.py:8]: This is the first time I'm hearing of the TF ranking module and probably explains why I couldn't find this metric when I first looked at tf.keras.metrics 🤦‍♂️ . Going to keep this implementation for now as its one less dependency for the project.

**Review by hmathlee — APPROVED** [review]
